In [1]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

In [2]:
import nltk
from nltk.corpus import stopwords
import string

In [3]:
# Download NLTK data
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to /Users/deven/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [4]:
# Load the dataset
url = "https://raw.githubusercontent.com/justmarkham/pycon-2016-tutorial/master/data/sms.tsv"
df = pd.read_csv(url, sep="\t", header=None, names=["label", "message"])

In [5]:
df

,label,message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will ü b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...


In [6]:
df.to_csv("spam_ham.csv")

In [7]:
# Text preprocessing function
def preprocess_text(text):
    # Lowercase the text
    text = text.lower()
    # Remove punctuation
    text = text.translate(str.maketrans("", "", string.punctuation))
    # Remove stopwords
    stop_words = set(stopwords.words("english"))
    # print(stop_words,len(stop_words))
    text = " ".join(word for word in text.split() if word not in stop_words)
    print(len(text))
    return text

In [8]:
# Apply preprocessing
df["message"] = df["message"].apply(preprocess_text)

82
23
135
35
43
88
46
113
121
109
80
110
115
111
11
111
15
51
40
136
20
35
30
68
34
93
21
63
48
40
65
95
15
35
96
70
32
16
21
87
51
65
124
6
44
25
36
26
52
104
28
89
62
167
100
44
123
28
12
26
56
42
16
36
87
101
74
135
40
20
28
14
34
24
6
24
42
24
12
31
20
9
17
9
13
33
71
96
28
36
49
119
120
99
91
122
30
26
109
24
41
53
51
113
128
33
65
22
64
66
25
20
15
73
118
109
92
114
59
60
129
115
34
131
26
36
10
127
4
74
15
7
31
21
96
147
23
9
28
107
51
16
16
119
18
14
32
123
57
18
24
24
45
19
113
208
16
16
65
96
112
56
35
39
112
117
41
112
55
46
41
39
71
17
73
40
41
18
87
30
74
96
11
34
10
22
89
19
109
24
63
50
114
27
60
49
26
19
41
70
14
15
62
120
23
18
52
92
27
55
25
46
16
16
13
17
45
35
34
17
74
20
104
20
90
108
46
110
29
124
17
80
23
26
38
118
29
38
17
100
100
39
42
40
125
30
37
13
27
81
137
40
141
8
98
30
54
29
10
113
39
3
60
33
98
12
22
30
135
26
94
35
103
118
31
19
5
24
19
19
39
15
84
18
4
82
37
2
95
75
125
24
40
14
9
42
122
42
90
38
50
64
102
39
16
112
19
19
56
105
27
141
107
68
91
18
22

In [9]:
df["message"].head()

0    go jurong point crazy available bugis n great ...
1                              ok lar joking wif u oni
2    free entry 2 wkly comp win fa cup final tkts 2...
3                  u dun say early hor u c already say
4          nah dont think goes usf lives around though
Name: message, dtype: object

In [10]:
vec = CountVectorizer(stop_words="english")

In [11]:
# Feature extraction using TF-IDF
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(df["message"])
y = df["label"]

In [12]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [13]:
# Train Naive Bayes classifier
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

MultinomialNB()

In [14]:
# Make predictions
y_pred = nb_classifier.predict(X_test)

In [15]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")

Accuracy: 0.9704035874439462
Classification Report:
              precision    recall  f1-score   support

         ham       0.97      1.00      0.98       966
        spam       1.00      0.78      0.88       149

    accuracy                           0.97      1115
   macro avg       0.98      0.89      0.93      1115
weighted avg       0.97      0.97      0.97      1115



d### Accuracy
The overall accuracy of the model is approximately 97.04%. This means that out of all the SMS messages in the test set, about 97.04% were correctly classified as either "ham" or "spam".

### Precision
Precision is the ratio of correctly predicted positive observations to the total predicted positives. It is an indicator of the accuracy of the positive predictions.
- For "ham": 0.97 (97%)
- For "spam": 1.00 (100%)

### Recall
Recall (or Sensitivity) is the ratio of correctly predicted positive observations to the all observations in actual class. It indicates how well the model can identify all positive instances.
- For "ham": 1.00 (100%)
- For "spam": 0.78 (78%)

### F1-Score
The F1-score is the harmonic mean of precision and recall. It provides a balance between the precision and recall, especially useful when you need to seek a balance between the two.
- For "ham": 0.98 (98%)
- For "spam": 0.88 (88%)

### Support: 
Support is the number of actual occurrences of the class in the dataset.
- For "ham": 966
- For "spam": 149

### Observations:
- Accuracy: 0.97 (97%), indicating the proportion of the total number of correct predictions.
- Macro Average:
  - Precision: 0.98 (98%)
  - Recall: 0.89 (89%)
  - F1-score: 0.93 (93%)
- Weighted Average:
  - Precision: 0.97 (97%)
  - Recall: 0.97 (97%)
  - F1-score: 0.97 (97%)
  
### Interpretation:
#### Ham Class:
- The model performs very well on the "ham" class with high precision (97%) and perfect recall (100%). This means it correctly identifies nearly all ham messages and rarely misclassifies spam messages as ham.
- The F1-score of 0.98 indicates a good balance between precision and recall.

#### Spam Class:
- The model has perfect precision (100%) for the "spam" class, meaning every message classified as spam is indeed spam. However, the recall for spam is 0.78, indicating that 22% of actual spam messages are not being identified (false negatives).
- The F1-score of 0.88 shows a good, but not perfect, balance between precision and recall for spam

In [24]:
df["message"]

0       go jurong point crazy available bugis n great ...
1                                 ok lar joking wif u oni
2       free entry 2 wkly comp win fa cup final tkts 2...
3                     u dun say early hor u c already say
4             nah dont think goes usf lives around though
                              ...                        
5567    2nd time tried 2 contact u u £750 pound prize ...
5568                          ü b going esplanade fr home
5569                          pity mood soany suggestions
5570    guy bitching acted like id interested buying s...
5571                                       rofl true name
Name: message, Length: 5572, dtype: object

In [28]:
df

,label,message
0,ham,go jurong point crazy available bugis n great ...
1,ham,ok lar joking wif u oni
2,spam,free entry 2 wkly comp win fa cup final tkts 2...
3,ham,u dun say early hor u c already say
4,ham,nah dont think goes usf lives around though
...,...,...
5567,spam,2nd time tried 2 contact u u £750 pound prize ...
5568,ham,ü b going esplanade fr home
5569,ham,pity mood soany suggestions
5570,ham,guy bitching acted like id interested buying s...


In [29]:
df["label"].value_counts()

label
ham     4825
spam     747
Name: count, dtype: int64

In [30]:
df["label_mapped"] = df["label"].map({"ham": 1, "spam": 0})

In [31]:
X, y = df["message"], df["label_mapped"]

In [34]:
X_train1, X_test1, y_train1, y_test1 = train_test_split(
    X, y, train_size=0.8, test_size=0.2, random_state=100
)
vec.fit(X_train1, y_train1)

CountVectorizer(stop_words='english')

In [35]:
vec.vocabulary_

{'dahow': 2247,
 'page': 5327,
 'want': 7743,
 'beautiful': 1350,
 'girl': 3261,
 'ive': 3947,
 'seen': 6297,
 'baby': 1259,
 'come': 2008,
 'common': 2020,
 'room': 6118,
 'sleepingand': 6530,
 'surfing': 6956,
 'maybe': 4629,
 'westshore': 7843,
 'hyde': 3727,
 'park': 5369,
 'village': 7655,
 'place': 5513,
 'near': 4968,
 'house': 3664,
 'hi': 3563,
 'darlin': 2263,
 'work': 7951,
 'trouble': 7390,
 'ijust': 3761,
 'talked': 7023,
 'mum': 4893,
 'morning': 4839,
 'really': 5897,
 'good': 3309,
 'time': 7229,
 'night': 5026,
 'im': 3771,
 'goin': 3294,
 'soon': 6633,
 'sad': 6175,
 'puppy': 5787,
 'noise': 5065,
 'hows': 3671,
 'champ': 1815,
 'leaving': 4274,
 'glasgow': 3273,
 'sorry': 6644,
 'ill': 3767,
 'later': 4241,
 'ree': 5940,
 'entry': 2739,
 'weekly': 7811,
 'comp': 2022,
 'chance': 1818,
 'win': 7885,
 'ipod': 3896,
 'txt': 7444,
 'pod': 5566,
 '80182': 723,
 'std': 6785,
 'rate': 5862,
 'tcs': 7056,
 'apply': 1100,
 '08452810073': 62,
 'details': 2395,
 '18': 347,
 'ho

In [36]:
len(vec.vocabulary_)

8180

In [44]:
X_train1.head()

2747                               k dahow many page want
4779    u r beautiful girl ive ever seen u r baby come...
338                                   sleepingand surfing
507     maybe westshore hyde park village place near h...
3312    hi darlin work u get trouble ijust talked mum ...
Name: message, dtype: object

In [47]:
dir(X_train)

['T',
 '__abs__',
 '__add__',
 '__array_priority__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__div__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__idiv__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__pow__',
 '__radd__',
 '__rdiv__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmatmul__',
 '__rmul__',
 '__round__',
 '__rsub__',
 '__rtruediv__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__weakref__',
 '_add_dense',
 '_add_sparse',
 '_allow_nd',
 '_arg_min_or_max',
 '_arg_min_or_max_axis',
 '_ascontainer',
 '_asfptype',
 '_asindices',
 '_binopt',
 '_broadcast_to',
 '_bsr_container',
 '_container',
 '_coo_containe

In [51]:
X_train1

2747                               k dahow many page want
4779    u r beautiful girl ive ever seen u r baby come...
338                                   sleepingand surfing
507     maybe westshore hyde park village place near h...
3312    hi darlin work u get trouble ijust talked mum ...
                              ...                        
4149    please call amanda regard renewing upgrading c...
1890                           gotta collect da car 6 lei
350      checking really miss seeing jeremiah great month
79                        still looking job much tas earn
3927                               havent add ü yet right
Name: message, Length: 4457, dtype: object

In [52]:
X_train1_transformed = vec.transform(X_train1)
X_test1_transformed = vec.transform(X_test1)

In [57]:
mnb = MultinomialNB()
mnb.fit(X_train1_transformed, y_train1)

y_test_pred = mnb.predict(X_test1_transformed)
accuracy1 = accuracy_score(y_test1, y_test_pred)

In [58]:
print(accuracy1)

0.9829596412556054


In [59]:
report1 = classification_report(y_test1, y_test_pred)

In [55]:
print(accuracy)

0.9704035874439462


In [60]:
print(report)

              precision    recall  f1-score   support

         ham       0.97      1.00      0.98       966
        spam       1.00      0.78      0.88       149

    accuracy                           0.97      1115
   macro avg       0.98      0.89      0.93      1115
weighted avg       0.97      0.97      0.97      1115



In [61]:
print(report1)

              precision    recall  f1-score   support

           0       0.97      0.90      0.93       147
           1       0.98      1.00      0.99       968

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



In [62]:
vectorizer.get_feature_names_out()

array(['008704050406', '0089my', '0121', ..., 'zyada', 'üll', '〨ud'],
      shape=(9437,), dtype=object)

In [64]:
vec2 = TfidfVectorizer()

### Trying to use TFID Vectorizer

In [65]:
X1 = vec2.fit(df["message"])
y1 = df["label"]

In [66]:
X1

TfidfVectorizer()

In [67]:
vec2.get_feature_names_out()

array(['008704050406', '0089my', '0121', ..., 'zyada', 'üll', '〨ud'],
      shape=(9437,), dtype=object)

In [68]:
dir(vec2)

['_CountVectorizer__metadata_request__fit',
 '_CountVectorizer__metadata_request__transform',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__sklearn_tags__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_build_request_for_signature',
 '_char_ngrams',
 '_char_wb_ngrams',
 '_check_feature_names',
 '_check_n_features',
 '_check_params',
 '_check_stop_words_consistency',
 '_check_vocabulary',
 '_count_vocab',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_tags',
 '_limit_features',
 '_more_tags',
 '_parameter_